spark/src/main/java/org/apache/iceberg/RewriteManifestsAction.java

```diff
@@ -0,0 +1,490 @@
+  private final Encoder<Entry> entryEncoder = Encoders.javaSerialization(Entry.class);
+  private final Encoder<Bin> binEncoder = Encoders.bean(Bin.class);
+
+  private Predicate<ManifestFile> predicate = manifest -> true;
+  private String stagingLocation = null;
+
+  RewriteManifestsAction(SparkSession spark, Table table) {
+    this.spark = spark;
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.specs = table.specs();
+    this.summary = Maps.newHashMap();
+    this.defaultParallelism = Integer.parseInt(
+        spark.conf().get("spark.default.parallelism", "200"));
+    this.snapshotIdInheritanceEnabled = PropertyUtil.propertyAsBoolean(
+        table.properties(),
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED,
+        TableProperties.SNAPSHOT_ID_INHERITANCE_ENABLED_DEFAULT);
+    this.targetManifestSizeBytes = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES,
+        TableProperties.MANIFEST_TARGET_SIZE_BYTES_DEFAULT);
+
+    if (table.io() instanceof HadoopFileIO) {
+      // we need to use Spark's SerializableConfiguration to avoid issues with Kryo serialization
+      SerializableConfiguration conf = new SerializableConfiguration(((HadoopFileIO) table.io()).conf());
+      fileIO = new HadoopFileIO(conf::value);
+    } else {
+      fileIO = table.io();
+    }
+  }
+
+  public RewriteManifestsAction rewriteIf(Predicate<ManifestFile> newPredicate) {
+    this.predicate = newPredicate;
+    return this;
+  }
+
+  public RewriteManifestsAction stagingLocation(String newStagingLocation) {
+    this.stagingLocation = newStagingLocation;
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsAction set(String property, String value) {
+    summary.put(property, value);
+    return this;
+  }
+
+  @Override
+  public RewriteManifestsActionResult execute() {
+    Preconditions.checkArgument(stagingLocation != null, "Staging location must be set");
+
+    List<ManifestFile> matchingManifests = findMatchingManifests();
+    if (matchingManifests.isEmpty()) {
+      return null;
+    }
+
+    Broadcast<FileIO> io = sparkContext.broadcast(fileIO);
+
+    int parallelism = Math.min(matchingManifests.size(), defaultParallelism);
+    JavaRDD<ManifestFile> manifestRDD = sparkContext.parallelize(matchingManifests, parallelism);
+    Dataset<ManifestFile> manifestDS = spark.createDataset(manifestRDD.rdd(), manifestEncoder);
+    Dataset<Entry> manifestEntryDS = manifestDS.flatMap(toEntries(io, specs), entryEncoder);
+
+    try {
+      manifestEntryDS.cache();
+
+      long manifestEntrySizeBytes = computeManifestEntrySizeBytes(matchingManifests);
+      Map<Integer, List<PartitionMetadata>> metadataSizeSummary = computeMetadataSizeSummary(
+          manifestEntryDS,
+          manifestEntrySizeBytes);
+
+      Map<Integer, Map<StructLike, Integer>> bins = computeBins(metadataSizeSummary);
```

Comment:
I'm not sure that computing bins is necessary. This handles skew so that the stage runtime is reasonable, but that doesn't matter for the outputs because `ManifestsWriter` will keep manifests to a reasonable size. So this is a runtime optimization only.

Also, because we're using a constant times the number of entries, `manifestEntrySizeBytes`, this is basically identical to using a count of entries and dividing the target size by the constant: `targetEntries = targetManifestSizeBytes / manifestEntrySizeBytes`. That means we're aiming for a some number of entries in each bin, which could be thought of as aiming for some number of bins given a total number of entries.

Since we're trying to divide the input data into some number of bins, cluster by partition, and handle skew, that's something that Spark can do for us automatically. By adding a global sort and a target number of tasks, we can get the same thing without doing our own bin packing calculation.

Reply:
One point to consider is about the precision we can get. Users most likely won’t play around with the number of samples that Spark will use to determine ranges for range partitioning. As a consequence, this can lead to a worse precision which is essential in this case. `ManifestsWriter`, introduced in this commit, is currently used only in cases when one partition has too much data for one partition. If we rely on Spark alone, we can get tasks that are let’s say 5 MB while we want to write manifests with 4 MB. `ManifestsWriter` will only harm here as it would split 5 MB into 4 MB and 1 MB files. At the same time, if we get rid of `ManifestsWriter` and write all data in a task to one manifest, we risk having large manifests. For example, if one task gets 8MB of metadata, it will produce one large manifest even though we wanted to have manifests of 4 MB. Moreover, the ranges might not be precise and a couple of partitions can be put together. All of that will degrade the performance and job planning might take more time.

Reply:
I'm not sure that the other approach would do much better on precision since the size of DataFile can vary and both ways assume that it is constant. And if we need manifests to be under 4MB, then I think the right way to do that is to set the size limit to 4MB and not to rely on packing.

I'd probably go with the simpler approach until we know that it doesn't achieve the precision necessary. The sort sampling defaults work for us almost all the time, and this is a small dataset because it is just metadata.

Reply:
I ran a quick test using the code I just pushed which leverages Spark for bin packing on 500 and 2600 manifests with a target size of 4 MB per manifest. As a result, I got 450 manifests where 25% of them were in 1-2 MB range and 25% were greater than 4 MB. The size of metadata per entry was pretty stable in this example. Increasing the number of samples did not help much. With manual bin-packing, I did not have any manifests larger than 4MB and most of them were bigger than 3 MB. I'll think about this more. I'll try to prototype how manual bin-packing can look like with the new approach and how it will impact the overall performance.

Reply:
Thanks for running the test. I think I'd prefer the simpler approach for maintainability, but I'm okay with either one going in.

Reply:
I went for automatic bin-packing using Spark but I have an additional step where I collect all manifest entries I got in a task before writing manifests. This means we will iterate through entries twice but it allows us to split the metadata more evenly.